presto-docs/src/main/sphinx/develop/presto-page.rst

```diff
@@ -0,0 +1,233 @@
+* Null flags
+* Total number of bytes in all values - 4 bytes
+* Concatenated values
+
+Let’s again take the example from the Null Flags section and say that we
+have a string column with 10 rows with nulls in zero-based rows 1, 4, 6, 7, 9.
+The non-null rows will have values: 0 - Denali, 2 - Reinier, 3 - Whitney,
+5 - Bona, 8 - Bear. We’ll have 4 bytes storing the number of rows: 10, followed
+by 40 bytes of offsets, followed by 3 bytes of null flags, followed by 1 bytes
+storing total size of all strings: 28, followed by the concatenated string values.
+Notice that we have offsets for all rows, not just the non-null rows.
+
+.. image:: presto-page-string-column.png
+  :width: 800
+
+ARRAY Encoding
+--------------
+
+* Elements column
+* Number of rows - 4 bytes
+* Offsets - (number of rows + 1) * 4 bytes; 4 bytes per offset
+* Null flags
+
+An array column with 10 rows is represented like this:
+
+.. image:: presto-page-array-column.png
+  :width: 400
+
+MAP Encoding
+------------
+
+* Keys column
+* Values column
+* Hash table size - 4 bytes
+* [optional] Hash table
+* Number of rows - 4 bytes
+* Offsets - (number of rows + 1) * 4 bytes; 4 bytes per offset
+* Null flags
+
+A map column with 10 rows is represented like this:
+
+.. image:: presto-page-map-column.png
+  :width: 600
+
+ROW Encoding
+------------
+
+* Number of fields - 4 bytes
+* One column per field
+* Number of rows - 4 bytes
+* Offsets - (number of rows + 1) * 4 bytes; 4 bytes per offset
+* Null flags
+
+Nested columns are serialized only for non-null rows. In the presence of
+null rows, the row numbers of the nested columns don’t match the top-level
+row numbers. Offsets are specifying the row numbers of the nested columns.
+
+Let’s again take the example from the Null Flags section and say that we
+have a column of type ROW(a, b, c, d) with 10 rows with nulls in zero-based
+rows 1, 4, 6, 7, 9. Nested columns will have only 5 rows and offsets will
+be: 0, 0, 1, 2, 0, 3, 0, 0, 4, 0. Offsets for null rows are zero.
+
+Note: Offsets are redundant information as they can be reconstructed
+from the null flags.
+
+.. image:: presto-page-row-column.png
+  :width: 600
+
+Additional Usage
+----------------
+
+SerializedPage format is also used to specify constant values in the plan
+fragment sent from the coordinator to the workers. In this case, the binary
+representation is converted to an ASCII string using base64 encoding.
+
+For example, the plan for the SELECT array[1, 23, 456] query includes a Project
+node with array[1, 23, 456] value represented as base64 encoded binary of the
+SerializedPage format.
+
+.. code-block:: text
+
+    - Project[projectLocality = LOCAL] => [expr:array(integer)]
+            Estimates: {rows: 1 (51B), cpu: 51.00, memory: 0.00, network: 0.00}
+            expr := [Block: position count: 3; size: 92 bytes]
+
+Also, this format is used to store intermediate data
+exchange_materialization_strategy is 'ALL'.
```

Comment:
when session property `exchange_materialization_strategy` is `ALL` and `temporary_table_storage_format` is `PAGEFILE`.